ragnarok_formats/src/map.rs

```diff
@@ -56,6 +74,54 @@ bitflags::bitflags! {
     }
 }
 
+impl FromBytes for QuadTreeData  {
+    fn from_bytes<Meta>(byte_stream: &mut ByteStream<Meta>) -> ConversionResult<Self> {
+
+        let mut stack = VecDeque::<(i32, i32)>::new();
+        let mut quadtree_vec: Vec<QuadTreeData> = Vec::new();
+        let mut quadtree_parent: Vec<i32> = Vec::new();
+        let mut child_index: i32 = 0;
+
+        // simulate a dfs using a stack
+        stack.push_back((0, 0));
+        // the node 0 doesn't have parent
+        quadtree_parent.push(-1);
+        while let Some((node,depth)) = stack.pop_front(){
+            let max = <[f32;3]>::from_bytes(byte_stream).trace::<Self>()?;
+            let min = <[f32;3]>::from_bytes(byte_stream).trace::<Self>()?;
+            let half_size = <[f32;3]>::from_bytes(byte_stream).trace::<Self>()?;
+            let center = <[f32;3]>::from_bytes(byte_stream).trace::<Self>()?;
+            let child: Vec<QuadTreeData> = Vec::new();
+            
+            quadtree_vec.push(QuadTreeData{
+                max,
+                min,
+                half_size,
+                center,
+                child,
+            });
+            if depth < 5 {
+                for _counter in 0..4 {
+                    child_index+=1;
+                    quadtree_parent.push(node);
+                    stack.push_back((child_index, depth+1));
+                }
+            }
+        }
+        // go from back to start to set the quadtree children in correct order
+        for index in (1..child_index as usize).rev() {
+            let quadtree = quadtree_vec.pop().unwrap();
+            quadtree_vec[quadtree_parent[index] as usize].child.push(quadtree);
+            if quadtree_vec[quadtree_parent[index] as usize].child.len() == 4 {
+                quadtree_vec[quadtree_parent[index] as usize].child.reverse();
+            }
+        }
+        quadtree_vec[0].child.reverse();
+        Ok(quadtree_vec.first().unwrap().clone())
```

Comment:
I was having some trouble understanding what is happening here, so I reworked this function a bit:

```rust
const MAX_DEPTH: usize = 5;
const CHILD_COUNT: usize = 4;

// Simulate a DFS using a stack.
let mut stack = VecDeque::<(usize, usize)>::from([(0, 0)]);
let mut nodes: Vec<QuadTreeData> = Vec::new();
let mut node_parents: Vec<usize> = Vec::new();

while let Some((parent_node, depth)) = stack.pop_front() {
    let max = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
    let min = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
    let half_size = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
    let center = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
    let children: Vec<QuadTreeData> = match depth < MAX_DEPTH {
        true => Vec::with_capacity(CHILD_COUNT),
        false => Vec::new(),
    };

    nodes.push(QuadTreeData {
        max,
        min,
        half_size,
        center,
        children,
    });

    if depth < MAX_DEPTH {
        for _counter in 0..CHILD_COUNT {
            node_parents.push(parent_node);
            let self_index = node_parents.len();
            stack.push_back((self_index, depth + 1));
        }
    }
}

// Iterate nodes backwards and assign them the their parents.
for parent_index in node_parents.into_iter().rev() {
    let child = nodes.pop().unwrap();
    let parent = &mut nodes[parent_index];

    parent.children.insert(0, child);
}

assert_eq!(nodes.len(), 1);
nodes.pop().unwrap()
```

Notable changes are:
- Adding constants for magic values (`MAX_DEPTH`, `CHILD_COUNT`)
- Renaming some variables to better express what they do
- Using `FromBytes::from_bytes` to let the Rust type system figure out the exact type
- Allocating vectors with the correct size to avoid reallocation
- Removed `child_index` since it can always be derived from the length of the `quadtree_parent`/`node_parents`
- Use `insert(0, ..)` for the children instead of `push(..)` followed by a `reverse`
- Not cloning the root node since we can just remove it from the vector

#### Questions

I also believe there were two errors in the logic of the original function.

1: You skip index `0` in line `112` but you also never iterate over `child_index`, only `child_index - 1` (because of the non-inclusive `..`) which left 2 nodes in the `quadtree_vec` at the end of the loop. If I understand correctly, there should only be 1 left (the root node).

2: The call to `quadtree_vec[0].child.reverse();` in line `119` reverses the ordering of child nodes a second time, since the first `reverse` is already happening in the loop above. I assume the root node should have the same ordering as the other nodes, so this line is incorrect I believe. Please correct me if I'm wrong!

Reply:
The first part of the algorithm is trying to construct the nodes from the QuadTree by generating recursively the nodes by a DFS using an stack. 

The second part of algorithm after the DFS, we construct the child relationship of the nodes from QuadTree by using the information about the parent that are gathered through the DFS. The way that I construct the child relationship is by running in the reverse order of DFS transversal order, you can do this by just following in the reverse order of how the nodes are being allocated in `quadtree_vec`, and assign the children to the correct parent.

1: For the Question 1, I forgot to iterate in `child_index` and should have used range equal (..=). Need to fix it.

2: For the Question 2, the second one I think I am correct, but I was using `reverse` whenever find out that the parent has four child, this means the child is well constructed and we can apply the `reverse`. And this condition was not being check in the root node, so I need to apply particularly to this case.

I need to create a test to check if the code is correct, maybe I am slipping in more stuff there.

Reply:
> And this condition was not being check in the root node

Why would it not be checked? The root node will also have four children after the loop right? Unless it didn't because of the first error, which is very possible.

Reply:
In the way that you have implemented the second part, the check is not needed anymore.

Oh, and I also think I am doing stack wrong should be `stack.push_front((self_index, depth + 1));` instead of `stack.push_back((self_index, depth + 1));`. I have an C++ bad muscle memory of typing because of vectors, .push_back()...
 I am going to push your modification and I would try to create a test, but it is not easy to create a test for this and to check if is everything all right. 
I will create another branch for the death/ reviving part.

Reply:
Very nice, thanks :)